sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteNonAggregateFirst.scala

```diff
@@ -0,0 +1,72 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.optimizer
+
+import org.apache.spark.sql.catalyst.expressions.{Alias, Literal}
+import org.apache.spark.sql.catalyst.expressions.aggregate.{AggregateExpression, First}
+import org.apache.spark.sql.catalyst.plans.logical._
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.catalyst.trees.TreePattern.FIRST
+import org.apache.spark.sql.types.IntegerType
+
+/**
+ * Rewrite aggregate plan with a single [[First]] function when grouping is absent. In such a case
+ * the query is equivalent to simple projection with limit 1.
+ *
+ * Input Pseudo-Query:
+ * {{{
+ *   SELECT FIRST(col) FROM table
+ * }}}
+ *
+ * Rewritten Query:
+ * {{{
+ *   SELECT col FROM table LIMIT 1
```

Comment:
Why we want this optimization? Is the purpose to implement one less function? Are there other similar aggregation function that can apply the same idea?

My concern is adding a rule to optimization process will increase the complexity of SQL compiler engine if the benefit is only to save implementation of a single aggregate function.  Conceptually having a aggregate function implemented  is much simpler.

Reply:
Another way to think that is the rule will be triggered on every query while only a very small set of query will gain the benefit?

Reply:
Good question, yes, this optimization only covers cases of one or more `FIRST` aggregate functions and only when `IGNORE NULLS` is not present, and no other aggregate functions are present. For other cases of the `FIRST` aggregate function, we plan to update the query execution to stop consuming input values after receiving the first one when possible (the current implementation consumes all the remaining values and discards them).

The advantage of generating a logical `LIMIT 1` in this case is that the optimizer can recursively push it down. This can result in more efficient execution since e.g. data source scans can prune partitions or do other improvements outside of the aggregation operator itself.

Re: complexity tradeoff vs. performance improvement, please leave opinions if needed.

Reply:
Thanks for the insight about potential gain on the performance improvement!

I see the tradeoff better now. I don't have strong opinion which one is better. Some systems are more sensible for query compilation latency.

For the common workload that Spark SQL deals with, I believe it does not care about compilation latency much but more favor of faster execution on large data set.

Reply:
Yeah, the optimization triggers in a small subset of cases. So we don't want to add a lot of complexity or compilation time. With the ideas suggested in this review, I think we can simplify the code for the rule implementation. And the pruning takes place in O(1) with bitmaps so it should not impose any compilation time unless the query actually contains FIRST aggregate function(s), and even then, very little.

Reply:
Yeah, the optimization triggers in a small subset of cases. So we don't want to add a lot of complexity or compilation time. With the ideas suggested in this review, I think we can simplify the code for the rule implementation. And the pruning takes place in O(1) with bitmaps so it should not impose any compilation time unless the query actually contains FIRST aggregate function(s), and even then, very little.

Reply:
I didn't know there is a way to do O(1) pruning. In that case this is not a concern.